### ISSUE TITLE
Consul Agent Does Not Register Services When leave_on_terminate Config Is Set To False

### ISSUE THREAD

**Issue by thaniri** [hashicorp/consul#5924]
#### Overview of the Issue

I have my consul agents using the -config-dir command line flag to read all configurations from a directory. They don't appear to read the service definition files when the leave_on_terminate config is set to false.

If I set it to true, the services will sync.

This flag is false by default on servers, and has no effect.

#### Reproduction Steps

Steps to reproduce this issue, eg:

1. Have a cluster up.
2. Get a client to join with a service definition as a json file in a configuration directory. Ensure that the consul configuration of said client has **leave_on_terminate** set to false. (Two config files, one for the consul config, one for the service definition, both will be read by the -config-dir command line flag)
3. Marvel as consul will totally ignore the service definition files. (No log message saying service has been synced, running a `consul catalog nodes -service=service` command will return nothing)

#### On Servers:

```
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 2
build:
        prerelease = 
        revision = c97c712e
        version = 1.4.2
consul:
        acl = enabled
        bootstrap = false
        known_datacenters = 4
        leader = true
        leader_addr = <hidden>
        server = true
raft:
        applied_index = 6972495
        commit_index = 6972495
        fsm_pending = 0
        last_contact = 0
        last_log_index = 6972495
        last_log_term = 8818
        last_snapshot_index = 6964811
        last_snapshot_term = 8818
        latest_configuration = [{Suffrage:Voter ID:<hidden> Address:<hidden>} {Suffrage:Voter ID:<hidden> Address:<hidden>} {Suffrage:Voter ID:<hidden> Address:<hidden>}]
        latest_configuration_index = 6837731
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Leader
        term = 8818
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 192
        max_procs = 2
        os = linux
        version = go1.11.4
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 89
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 6172
        members = 108
        query_queue = 0
        query_time = 5
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1410
        members = 12
        query_queue = 0
        query_time = 1
```

#### OS
Ubuntu 16.04.6 LTS

#### Systemd Unit File
I'm running consul as a systemd service.
```
[Unit]
Description=Consul Agent
After=network.target

[Service]
User=consul
Group=consul
ExecStart=/usr/local/bin/consul agent -config-dir/etc/consul -log-file=/var/log/consul/consul.log
ExecReload=/bin/kill -HUP $MAINPID
Restart=on-failure
LimitNOFILE=131072

[Install]
WantedBy=multi-user.target
```

#### Client Config

```
{
    "acl": {
        "default_policy": "deny",
        "down_policy": "extend-cache",
        "enabled": true,
        "tokens": {
            "agent": "<hidden>"
        }
    },
    "addresses": {
        "http": "<hidden> 127.0.0.1"
    },
    "bind_addr": "<hidden>",
    "bootstrap_expect": 0,
    "ca_file": "<hidden>",
    "cert_file": "<hidden>",
    "data_dir": "<hidden>",
    "datacenter": "<hidden>",
    "disable_remote_exec": true,
    "disable_update_check": true,
    "encrypt": "<hidden>",
    "key_file": "<hidden>",
    **"leave_on_terminate": false,**
    "log_level": "info",
    "node_name": "<hidden>",
    "rejoin_after_leave": true,
    "retry_interval": "60s",
    "retry_join": [
        "<hidden>",
        "<hidden>",
        "<hidden>"
    ],
    "server": false,
    "verify_incoming": true,
    "verify_outgoing": true,
    "verify_server_hostname": true
}
```

#### Service Definition

```
{
"service": {
"checks": [],
"enable_tag_override": false,
"id": <hidden>,
"meta" {
 <hidden>: <hidden>
},
"name": <hidden>,
"port": <hidden>
"tags" [
  <hidden>
],
"token": <hidden>
}
```

### Log Fragments

Logs are totally silent in regards to service registry when running a client with the `leave_on_terminate` flag set to false.

With `leave_on_terminate` set to true, I'll see a log on service start like `"Service "exporter-node" in sync`

### Edits because I kept digging:

* This is only happening on VMWare virtual machines and baremetal servers running Consul in client mode. Neither Xen nor KVM virtual machines in AWS are having this issue. Consul configs are identical, as it's all managed by a config management tool. Only difference is stuff like node names and which servers to retry_join to.
* If you set the terminate_on_leave flag to false, restart consul. Delete that line from config, restart consul again. The service will not register. One one of my clients if I explicitly set it to true the service will register. On another it won't still... Crisis of reproducibility. The vast majority of clients are having this problem, but I can find a client that doesn't.
* curl commands to the API hang in addition to using consul command line

**Comment by thaniri** [comment]
Bonus points. If you set the terminate_on_leave flag to false, restart consul. Set it to true, restart consul again. The service will not register.

**Comment by thaniri** [comment]
I've discovered the consul debug command. Here are stream of consciousness notes on what interesting things I find.

* In the clusters with problems there `consul.memberlist.probeNode` has a mean value of 834. In the clusters without problems this is 2.
* In the clusters with problems, there are metrics for autopilot and WAN configuration that we thought we got rid of. But apparently not.
* New clients being brought up don't have this problem (with the leave_on_terminate flag set to false on first attempt to start consul). It appears to only be servers that have been standing for a long time are refusing to join the consul cluster.
